File: configuration/src/main/java/com/synopsys/integration/configuration/parse/ListValueParser.java
package com.synopsys.integration.configuration.parse;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

public class ListValueParser<T> extends ValueParser<List<T>> {
    @NotNull
    private final ValueParser<T> valueParser;

    @NotNull
    private final String delimiter;

    public ListValueParser(@NotNull ValueParser<T> valueParser) {
        this(valueParser, ",");
    }

    public ListValueParser(@NotNull ValueParser<T> valueParser, @NotNull String delimiter) {
        this.valueParser = valueParser;
        this.delimiter = delimiter;
    }

    @NotNull
    @Override
    public List<T> parse(@NotNull String rawValue) throws ValueParseException {
        List<T> parsedValues = new ArrayList<>();

        for (String element : StringUtils.splitPreserveAllTokens(rawValue, delimiter)) {
            String trimmedElement = element.trim();

            if (StringUtils.isBlank(trimmedElement)) {
                throw new ValueParseException(rawValue, "List",
                    String.format(
                        "Failed to parse list '%s'. The list must be comma separated and each element in the list must not be empty (at least one character that is not whitespace).",
                        rawValue
                    )
                );
            } else {
                parsedValues.add(valueParser.parse(trimmedElement));
            }
        }

        return parsedValues;
    }
}

File: configuration/src/main/java/com/synopsys/integration/configuration/parse/ListValueParser.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.parse

/**
 * Splits the configuration provided value to a list of T around occurrences of the specified [delimiters].
 *
 * @param valueParser The ValueParser to be applied to each entry in the list.
 * @param delimiters One or more strings to be used as delimiters. Defaults to comma-separated.
 *
 * To avoid ambiguous results when strings in [delimiters] have characters in common, this method proceeds from
 * the beginning to the end of this string, and matches at each position the first element in [delimiters]
 * that is equal to a delimiter in this instance at that position.
 */
open class ListValueParser<T>(private val valueParser: ValueParser<T>, private vararg val delimiters: String = arrayOf(",")) : ValueParser<List<T>>() {
    override fun parse(value: String): List<T> {
        return value.split(*delimiters)
                .map { valueParser.parse(it) }
                .toList()
    }
}